contracts/strategies/lift/LiftStrategy.sol

```diff
@@ -0,0 +1,271 @@
+    }
+    IERC20(underlying).safeTransfer(vault, amount);
+  }
+
+  /*
+  *   Note that we currently do not have a mechanism here to include the
+  *   amount of reward that is accrued.
+  */
+  function investedUnderlyingBalance() external view returns (uint256) {
+    if (address(rewardPool) == address(0)) {
+      return IERC20(underlying).balanceOf(address(this));
+    }
+    // Adding the amount locked in the reward pool and the amount that is somehow in this contract
+    // both are in the units of "underlying"
+    // The second part is needed because there is the emergency exit mechanism
+    // which would break the assumption that all the funds are always inside of the reward pool
+    return rewardPool.balanceOf(address(this)).add(IERC20(underlying).balanceOf(address(this)));
+  }
+
+  /*
+  *   Governance or Controller can claim coins that are somehow transferred into the contract
+  *   Note that they cannot come in take away coins that are used and defined in the strategy itself
+  *   Those are protected by the "unsalvagableTokens". To check, see where those are being flagged.
+  */
+  function salvage(address recipient, address token, uint256 amount) external onlyControllerOrGovernance {
+     // To make sure that governance cannot come in and take away the coins
+    require(!unsalvagableTokens[token], "token is defined as not salvagable");
+    IERC20(token).safeTransfer(recipient, amount);
+  }
+
+  /*
+  *   Get the reward, sell it in exchange for underlying, invest what you got.
+  *   It's not much, but it's honest work.
+  *
+  *   Note that although `onlyNotPausedInvesting` is not added here,
+  *   calling `investAllUnderlying()` affectively blocks the usage of `doHardWork`
+  *   when the investing is being paused by governance.
+  */
+  function doHardWork() external onlyNotPausedInvesting restricted {
+    uint256 len = stakes.length;
+    require(len<=200, "Too many stakes, withdraw first");
+    uint256 rewardBalanceBefore = boardRoom.getbalanceOfShare(address(this));
+    rewardPool.stakeInBoardroom();
+    uint256 rewardBalanceAfter = boardRoom.getbalanceOfShare(address(this));
+    uint256 rewardStaked = rewardBalanceAfter.sub(rewardBalanceBefore);
+    if (rewardStaked > 0) {
+      stakes.push([block.timestamp, rewardStaked]);
+    }
+    investAllUnderlying();
+  }
+
+  function _getReward() internal {
+    uint256 len = stakes.length;
+    require(len<=200, "Too many stakes, withdraw first");
+    uint256 rewardBalanceBefore = boardRoom.getbalanceOfShare(address(this));
+    rewardPool.stakeInBoardroom();
+    uint256 rewardBalanceAfter = boardRoom.getbalanceOfShare(address(this));
+    uint256 rewardStaked = rewardBalanceAfter.sub(rewardBalanceBefore);
+    if (rewardStaked > 0) {
+      stakes.push([block.timestamp, rewardStaked]);
+    }
+  }
+
+  function withdrawRewardShareOldest() external onlyMultiSigOrGovernance {
```

Comment:
There needs to be a way of withdrawing multiple oldest shares simultaneously. If the stakes keep piling up, it will become very costly to withdraw the shares for us.

Reply:
This is what I wanted to do from the start. It is impossible. The `Boardroom` [contract](https://etherscan.io/address/0x3223689b39db8a897a9a9f0907c8a75d42268787#code) has protections in place that one address only interact with 1 protected function per block. So it is impossible to remove multiple stakes at once. We can only withdraw 1 stake, or all stakes.